okhttp-ws-tests/src/test/java/com/squareup/okhttp/internal/ws/RealWebSocketTest.java

```diff
@@ -207,7 +219,92 @@
     }
   }
 
-  @Test public void serverCloseThenWritingThrows() throws IOException {
+  @Test public void clientWritingThrowsSendsClientClose() throws IOException {
+    final IOException brokenException = new IOException("Broken!");
+    RequestBody brokenBody = new RequestBody() {
+      @Override public MediaType contentType() {
+        return TEXT;
+      }
+
+      @Override public void writeTo(BufferedSink sink) throws IOException {
+        throw brokenException;
+      }
+    };
+
+    try {
+      client.sendMessage(brokenBody);
+      fail();
+    } catch (IOException e) {
+      assertSame(brokenException, e);
+    }
+
+    // A failed write prevents further use of the WebSocket instance.
+    try {
+      client.sendPing(new Buffer().writeUtf8("Ping!"));
+      fail();
+    } catch (IllegalStateException e) {
```

Comment:
... rule of thumb: if it's a programming bug, IllegalStateException. If it's not, IOException. Here I think it isn't a programming bug.